### PR TITLE
Docs: Add matcher examples

### LINKED ISSUE #41: Improve JSDoc
As the API reference will be auto-generated with [TypeDoc](https://typedoc.org/), we want to improve the JSDoc to add more information to the documentation. we want to add the following changes:

- [x] Add JSDoc to all assertion classes
- [x] Add `@param` to all generic types
- [ ] Add `@example` to all assertion matcher methods to show their usage in the documentation
- [x] Change `@link` of built-in data types to link to the official JavaScript documentation.
  - For instance, `{@link Error}` should point to the [MDN Error docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error)

### PR DESCRIPTION
**Pull request by neolight1010**
This PR resolves #41 . Adding examples to matchers of:
- [x] Assertion
- [x] ArrayAssertion
- [x] BooleanAssertion
- [x] FunctionAssertion
- [x] DateAssertion
- [x] ErrorAssertion
- [x] NumberAssertion
- [x] ObjectAssertion
- [x] PromiseAssertion
- [x] StringAssertion

### CODE REVIEW

File: src/lib/helpers/TypeFactories.ts [hunk 1]
```diff
@@ -95,7 +95,7 @@ export interface StaticTypeFactories {
    *   age: number;
    * }
    *
-   * Typefactories.object<User>(); // a `User` object factory
+   * TypeFactories.object<User>(); // a `User` object factory
```
Comment (JoseLion) [inline src/lib/helpers/TypeFactories.ts:98]: 👍

**Review by JoseLion — COMMENTED** [review]
Thank you so much for your contribution @NeoLight1010! Overall this PR looks great. I left a few suggestions and comments, nothing major though. Excellent job 🎉

**Comment by NeoLight1010** [comment]
Thank you for reviewing my PR @JoseLion ! I added a commit with your suggestions.

**Review by JoseLion — APPROVED** [review]
This is looking great! Just found one last missing backtick set. Thanks a lot for your help on this 🎉

File: src/lib/Assertion.ts [hunk 2]
```diff
@@ -300,6 +338,13 @@ export class Assertion<T> {
   /**
    * Check if the value is shallow equal to another value.
    *
+   * @example
+   * ```
+   * expect(3 + 2).toBeSimilar(5);
+   * expect({ a: 1 }).toBeSimilar({ a: 1 });
+   *
+   * expect({ a: { b: 1 } }).not.toBeSimilar({ a: {b: 1} });
+   *
```
[the remark this answers is not in the document]
Comment (NeoLight1010) [reply src/lib/Assertion.ts]: Whoops! Fixed, thank you!

**Pull request merged**